### ISSUE TITLE
"Error building AzureRM Client" on Azure Devops

### ISSUE THREAD

**Issue by mastoj** [pulumi/pulumi#4840]
Trying to deploy some azure servicebus things with the devops pipeline I've configured. I have configured a service principal and configured the `pulumi.dev.yaml` like such:

```
config:
  azure:clientId: XXXXXXX
  azure:clientSecret:
    secure: SECURESTRING
  azure:subscriptionId: GUIDAGAIN
  azure:tenantId: YYYY
```

The stage description that fails look like this:

```
- stage: DeployDev
  pool:
    vmImage: 'ubuntu-latest'
  jobs:
  - deployment: DeployToDev
    displayName: Deploy Dev
    # creates an environment if it doesn't exist
    environment: 'sweetspot-dev'
    strategy:
      runOnce:
        deploy:
          steps:
          - checkout: self
          - task: Bash@3
            displayName: Setup pulumi
            inputs:
              targetType: inline
              script: |
                curl -fsSL https://get.pulumi.com | bash
                export PATH=$PATH:$HOME/.pulumi/bin
                echo "##vso[task.setvariable variable=path]$PATH"
                pulumi login
                curl -sL https://aka.ms/InstallAzureCLIDeb | sudo bash
            env:
              PULUMI_ACCESS_TOKEN: $(pulumi.access.token)            
          - task: Bash@3
            displayName: Deploy dev
            inputs:
              targetType: inline
              script: 
                make deploy
```

When running on Azure Devops I get the following error:

> error: Error building AzureRM Client: Azure CLI Authorization Profile was not found. Please ensure the Azure CLI is installed and then log-in with `az login`.
 
The `make deploy` executes a F# fake script that sets the stack and call `pulumi up -y`. The pulumi code I have deploy a kubernetes app and now also some servicebus stuff. It did work fine when I deployed only the kubernetes stuff, but I guess that is because I don't need to run the az cli tool for that. What might I be doing wrong?

**Comment by praneetloke** [comment]
That's certainly strange. @mastoj are you able to post the contents of your `Makefile` and the Pulumi app itself here?

Also as a side note, we do have a Task Extension for Azure Pipelines available for free in the [VS Marketplace](https://marketplace.visualstudio.com/items?itemName=pulumi.build-and-release-task).  With the task you would be able to specify an [Azure Service Connection](https://docs.microsoft.com/en-us/azure/devops/pipelines/library/connect-to-azure?view=azure-devops) that you have setup in Azure DevOps. It makes managing the Azure Service Principal itself a bit easier and would allow you to remove the `azure:*` Service Principal config from your `Pulumi.dev.yaml`.

**Comment by mastoj** [comment]
@praneetloke, I made the sample application public so you can find everything here: https://github.com/mastoj/sweetspot

I've seen the Task Extensions, but I prefer to script everything so I'm not tied to any CI and then just use the CI/CD service to call the scripts. That also makes it easier to test thing locally I think.

**Comment by mastoj** [comment]
To be clear, everything works fine locally. I take for granted that when I have specified the azure things in my stack config that will be used for the azure calls, even if I'm logged in with a different user locally, is that correct?

**Comment by mastoj** [comment]
Also, I'm pretty sure it is some kind of user error most likely :)

**Comment by mastoj** [comment]
Just verified that I actually have `az` installed, and it is there.

**Comment by mastoj** [comment]
Did some more debugging on my end ran everything in a docker container locally (with basically the same commands I have in my pipeline config), and that works just fine. So for some reason it seems like the azure config isn't picked up when running on devops... not sure why that is though.

**Comment by praneetloke** [comment]
@mastoj I was able to use (most of) your Azure Pipelines configuration to deploy two of our example Azure apps by setting the Azure Service Principal credentials in the stack configuration. For good measure, I tried TS and F#, though, they shouldn't have any difference in how the provider credentials are read. In any case, I am going to clone your GH repo and take a look to see if I can reproduce your issue with that repo.

**Comment by mastoj** [comment]
https://github.com/mastoj/sweetspot/tree/minimal_error is a close to minimal way to reproduce the error.

**Comment by praneetloke** [comment]
I encountered some build errors (remove the now non-existent projects from the solution file, add Nuget deps for the project etc.) in your project that I had to fix, but have managed to deploy your project as-is for the most part. I added the Service Principal to the stack config. I didn't need to modify anything else about your project. Note that I am on a Mac so I used Visual Studio for Mac to load the solution and fixed the dotnet build errors that I was seeing in both Azure Pipelines, as well as in the IDE.

Here's a snippet of the build log:

```
2020-06-23T16:26:12.7517903Z Updating (dev):
2020-06-23T16:26:13.3256110Z     pulumi:pulumi:Stack Sweetspot.Infrastructure.Application-dev  running 'dotnet build -nologo .'
2020-06-23T16:26:13.9977266Z     pulumi:pulumi:Stack Sweetspot.Infrastructure.Application-dev    Determining projects to restore...
2020-06-23T16:26:14.3794158Z     pulumi:pulumi:Stack Sweetspot.Infrastructure.Application-dev    All projects are up-to-date for restore.
2020-06-23T16:26:14.6988219Z     pulumi:pulumi:Stack Sweetspot.Infrastructure.Application-dev    Sweetspot.Infrastructure.Application -> /home/vsts/work/1/s/src/infrastructure/Sweetspot.Infrastructure.Application/bin/Debug/netcoreapp3.1/Sweetspot.Infrastructure.Application.dll
2020-06-23T16:26:14.7085362Z     pulumi:pulumi:Stack Sweetspot.Infrastructure.Application-dev  
2020-06-23T16:26:14.7089250Z     pulumi:pulumi:Stack Sweetspot.Infrastructure.Application-dev      0 Warning(s)
2020-06-23T16:26:14.7099967Z     pulumi:pulumi:Stack Sweetspot.Infrastructure.Application-dev  
2020-06-23T16:26:14.7118543Z     pulumi:pulumi:Stack Sweetspot.Infrastructure.Application-dev  Time Elapsed 00:00:01.03
2020-06-23T16:26:15.1654223Z     pulumi:pulumi:Stack Sweetspot.Infrastructure.Application-dev  'dotnet build -nologo .' completed successfully
2020-06-23T16:26:16.5475312Z 
2020-06-23T16:26:18.5726765Z  +  pulumi:pulumi:Stack Sweetspot.Infrastructure.Application-dev creating 'dotnet build -nologo .' completed successfully
2020-06-23T16:26:21.7488465Z  +  azure:core:ResourceGroup deleteme creating 
2020-06-23T16:26:22.6865122Z  +  azure:core:ResourceGroup deleteme created 
2020-06-23T16:26:23.0984590Z  +  pulumi:pulumi:Stack Sweetspot.Infrastructure.Application-dev created 
2020-06-23T16:26:23.3492352Z  
2020-06-23T16:26:23.3492596Z Outputs:
2020-06-23T16:26:23.3492838Z     resourcegroupname: "deleteme4ed451e6"
2020-06-23T16:26:23.3493040Z 
2020-06-23T16:26:23.3493268Z Resources:
2020-06-23T16:26:23.3493451Z     + 2 created
```

So it does seem to work as intended. I am doubting if the Service Principal creds you are using has some problem that is causing the provider to throw a misleading error.

To be on the same page, here's what I did to create my Service Principal:

`az ad sp create-for-rbac --name ServicePrincipalName` (I replaced `ServicePrincipalName` with my own.)

Source: https://docs.microsoft.com/en-us/cli/azure/create-an-azure-service-principal-azure-cli?view=azure-cli-latest#password-based-authentication

**Comment by praneetloke** [comment]
I was able to successfully run the project with @mastoj's Azure SP creds. Moreover, when I sent him the project files I used, it worked for him too in Azure Pipelines. So there doesn't seem to be a problem with the structure of the project or any of the versions involved. I have asked @mastoj to try to use the original "broken" branch in GitHub Actions to see if the Az Pipelines behavior can be replicated.

**Comment by mastoj** [comment]
After playing around with this with @praneetloke for the last two nights I think we've narrowed it down to being an issue with wrong/inconsistent line endings. I changed to `git config --global core.autocrlf input` and also ran `find src/ -type f -exec dos2unix {} \;` and it sort of started working. I've been editing the files from windows as well as from wsl, so it might have been some inconsistencies around line endings causing this issue.

With that said, I thought line endings was a "solved" problem :), so I guess I will not be the last one hitting this issue... or maybe I am.

**Comment by mastoj** [comment]
Apparently I was wrong, or at least about the part that I have it working. It looked like it was working since the build was green, but reading through the logs still showed the same issue. The reason it was green was because @praneetloke added one line to the script that ran `pulumi destroy` after `pulumi up`, so even if `up` failed the `destroy` worked and returned 0 as exit code. I should have noticed that before. No I'm curious if @praneetloke really got it working with his setup or if he was fooled by this as well :)

**Comment by praneetloke** [comment]
> No I'm curious if @praneetloke really got it working with his setup or if he was fooled by this as well :)

Yeah this [comment](https://github.com/pulumi/pulumi/issues/4840#issuecomment-648280855) above has the logs from my pipeline execution.

Note that @mastoj has tried to start a fresh F# project and deployed that successfully from AzDO. So there is something specific to the original F# project that is still causing a problem and moreover only in his Az Pipeline build. I have not been able to reproduce it with either of our SP creds.

**Comment by praneetloke** [comment]
I am closing this per my discussion with @mastoj. Here's what we learned:

* I could never reproduce the issue even when I cloned the specific branches referenced in earlier comments. In fact, when I zipped-up the project and sent it to @mastoj he, too, could not reproduce it. Moreover, it was literally the same project that he had problems with, but when he overwrote them with my copy it all worked just fine.
* @mastoj is using WSL with VS Code on Windows and we think that a combination of bad file-endings in the `Pulumi.dev.yaml` may have rendered the file unreadable in the Az Pipelines build agent, which led to the Azure provider not being able to read the credentials in the stack config and therefore falling back to try Az CLI-based authentication.
* After @mastoj started a new project with exactly the same Pulumi app everything worked fine in Az Pipelines. Weirdly enough, the old project files where he encountered this issue still continues to be a problem for him but he has since moved on.

So if anyone encounters a similar issue, it is good to check your git commit settings and especially with WSL on VS Code.
